### ISSUE TITLE
Add a object-verb mode to have visual feedback before action

### ISSUE THREAD

**Issue by eugenesvk** [NeoVintageous/NeoVintageous#870]
To delete a word in vim, you 
  - start a delete command (`action`), then 
  - select what text (`object`) to delete

Most of other non-modal editors, and also a few modal ones (Helix, Kakoune), use a more intuitive paradigm:
  - you select text first and then 
  - perform an action

  so you get immediate __visual feedback__ __before__ doing anything, which is especially important in the case of __multiple cursors__ where reasoning about what objects your command will apply to is tricky

Could you please describe what kind of changes to the plugin would be required to implement it?
Does it need to run permanently in  visual mode/select mode with tweaking some commands to break selection to allow for text navigation (just like in Helix movement commands only select the last text object they jump over)?

The select mode seems like a good fit - it allows you to select multiple words and cycle over selections etc., but then it's blocked by the fact that multicursor plugin

> Yes. Multiple cursor mode uses select mode. I don't know if this is the case in Vim, but in nv it was easier to implement this way.

https://github.com/NeoVintageous/NeoVintageous/issues/865#issuecomment-1537055149

**Comment by gerardroche** [comment]
I'm afraid I don't understand what it is that you're trying to do. There is the highlighted yank plugin which highlights the text you yank e.g. press `y5j` will highlight 5 lines for 1000ms. You can configure it the time and style of the highlighting:

```

    // Plugin
    //
    // A port of https://github.com/machakann/vim-highlightedyank.
    "highlightedyank": true,

    // Assign number of time in milliseconds.
    "highlightedyank_duration": 1000,

    // Valid values are:
    //     fill
    //     outline
    //     underline
    //     squiggly_underline
    //     stippled_underline
    "highlightedyank_style": "fill",

```
 
I think want to the same highlighting *before* operations like delete and cut e.g. you want `d5j` to highlight the 5 lines before deleting?

**Comment by eugenesvk** [comment]
> think you want the same highlighting as yank except _before_ operations like delete e.g. you want `d5j` to highlight the 5 lines before deleting

This is still verb (delete) object (5 lines)

I want to highlight 5 lines first before I decide what to do with them, yank or delete (or realise that I wanted 1 more line and add it before deciding to delete), so the delete/yank operations would be invoked after the highlight is complete. Now, that's what the Visual mode is for, but then in V everything always extends, so this wouldn't be possible (using your Multiline select plugin):

https://github.com/NeoVintageous/NeoVintageous/assets/12956286/7338ddc9-2f3f-493b-be13-f16bc806710f

I can move across selections without always extending. But then the select mode only operates on matches

And for single word operations regular movement operations like `w` would preselect the word, so instead of `dw` you'd move with `w` and then `d` the already highlighted word. And moving to the beginning of the line would also preselect from the cursor position to the beginning (though these single-move highlights can be annoying to some).

So this would be like Visual mode without selection extensions, but also without deselecting everything, so in the multiline select video above exiting to this mode would retain the selections, and then you'd enter real Visual mode to extend

(if you're curious you might want to check out Helix or Kakoune for better visual demonstration, I realize it's a challenge to describe the paradigm with just the words)

**Comment by gerardroche** [comment]
Yes, I'm finding it difficult to understand it. You know that when you're in
multiple-cursor mode you can press `<Esc>` to exit to normal mode retaining all
the cursors and then use normal commands on all the cursors e.g. you can enter
Visual mode and select words `<Esc>vww`.

The commands would apply to all cursors, but if I'm understanding you correctly,
you want to exit the "last" cursor leaving all the other cursors as they are,
and then work on that last cursor i.e. "extending" it?

**Comment by gerardroche** [comment]
And not even just the "last" cursor. You want to be able do that on any of the
cursors i.e. you want to be able to work on each cursor individually,
essentially building a complex visual selection.

**Comment by eugenesvk** [comment]
> We would have to enter a Visual mode to keep selections

Ok, but then how to adjust this Visual mode NOT to extend selections on move. Basically, how to replace Normal mode with `Visual mode minus extend selection`. This way you always have something selected (even if it's just 1 char) and the delete commands will work by deleting right away instead of waiting for a selection

**Comment by gerardroche** [comment]
ahh, I think I understand now. In Vim you can move the cursor in Visual. You can't do this in Sublime Text. It's a known issue, I even opened an issue about it https://github.com/sublimehq/sublime_text/issues/3033. It's in the known issues section here https://github.com/NeoVintageous/NeoVintageous/issues/854.

That's what you're talking about isn't it?

**Comment by eugenesvk** [comment]
afaik this is about moving **within** a selection (linewise visual mode selects the current line, but moving horizontally within that line doesn't change the whole-line selection)

But then I was talking about moving the cursor in the regular Visual mode, where every movement extends the selection, which is undesirable, so compare it to the Visual-without-extension mode:

https://github.com/NeoVintageous/NeoVintageous/assets/12956286/51578da0-a075-4a28-8270-458a8aa21d81

or moving with `count`

https://github.com/NeoVintageous/NeoVintageous/assets/12956286/b22075f4-66ed-41c5-affa-e04e434f3089

and then if you want to delete all those 9 words, then you'd switch into the Visual-with-extension mode (basically the same as vim's Visual), so this part is identical to vim's `v9wx`. Or deleting a word you'd do `viwd` instead of `diw`, and again, this is possible already.

So you could say "just perma-stay in Visual mode", BUT then Visual mode has this issue where every move command extends selection unlike see video #1 above

**Comment by gerardroche** [comment]
I kind of understand. I think it's too much of a shift from how Vim works. Are there any Vim plugins that try to implement this?

**Comment by eugenesvk** [comment]
I'm not aware of plugins, unfortunately, only seen some discussions re. how to achieve kakoune mode in Vim, but no implemented solution

By the way, as far as I understand it, it's further complicated by the fact that in the plugin modes are built into commands. So, for example, in Helix you have 
- `move with selection, but don't extend selection` command and 
- `move with selection and extend selection` command

and you could use either of these commands in any mode. But in NV the mode check is part of the single `nv_vi_w` command, so I can't bind in normal mode to `w` a command that also selects just like in Visual mode?

Or maybe there is a relatively simple way to "pass/override" the mode input `'mode': get_mode(view),` in each command so I could have `nmap w w mode=visual` and now in normal mode `w` would extend selection? What do you think?

**Comment by gerardroche** [comment]
There are number of ways it could be implemented. On the face of it I can't see any way to hack it other than adding new modes and adding the functionality to every motion and operator. 

I feel I have to warn you that I won't necessarily merge something like this and it seems like a large project to spend time on. It's not something that I'm interested in working on so I'm going to close the issue but feel free to ask questions if you're going to have a go at it anyways.

**Comment by eugenesvk** [comment]
> On the face of it I can't see any way to hack it other than adding new modes and adding the functionality to every motion and operator.

That's exactly what I wanted to avoid :) as it's investing a lot of time into broken foundations (the primitive commands should be mode-agnostic so you could use them anywhere), thus I thought you might know of some better hack

> I feel I have to warn you that I won't necessarily merge something like this and it seems like a large project to spend time on

Yeah I surmised that much after all the other conversations re fixing some broken vim designs
